### ISSUE TITLE
Add search to landing page docs

### ISSUE THREAD

**Issue by travis1111** [eclipse/codewind#1786]
We used to have search on mc site, we may want to have search function back on codewind site.

**Comment by sishida** [comment]
It'd be great to have search on the website!

The previous website, however, had significant limitations to search. For example, if I recall, if you searched for more than one word at a time, you had to enter the words in the same order as how they were listed in the keywords metadata part of the Markdown file.

It'd be ever better if we could remove some of these limitations when adding search. 😃

**Comment by sghung** [comment]
Che uses jekyll-search (which we also used in microclimate, but was bad). Maybe look into Google custom search or other third party stuff.

**Comment by jcockbain** [comment]
@sghung, @sishida I assume this is still something that would be good to have? I was looking through the website backlog and thought this might be good for @alif-munim or me to pick up.

**Comment by sghung** [comment]
Yes @jcockbain this is still something good to have. Even if it is the jekyll-search, which other people still use, it's still better than nothing

**Comment by alif-munim** [comment]
Hey all, I'm currently looking into lunr.js which uses client-side js to search through the site. I'll keep you posted.

**Comment by kimholmes** [comment]
Adding from Slack:
![image (3)](https://user-images.githubusercontent.com/21316329/82481425-45d2a180-9aa3-11ea-943f-92fd982285d2.png)
![image (4)](https://user-images.githubusercontent.com/21316329/82481456-5256fa00-9aa3-11ea-8ccd-2fd1e27f49f2.png)

**Comment by alif-munim** [comment]
Update: 
- Added search bar with magnifying glass icon to navigation
- Alternative design for search results using bootstrap cards

I'll get this all up on a new branch as soon as I've set up the GitHub workflow on my machine 👍 

![image](https://user-images.githubusercontent.com/59207653/82500588-a2dd5000-9ac1-11ea-848d-15841c6e5901.png)

**Comment by micgibso** [comment]
Tweaking the status in response to request from MelHopper.

**Comment by MelHopper** [comment]
Search results list different types of content - Docs, Guides, Blogs....Guides and Blog cards contain description of the content. For the Docs cards,  can we display the beginning text of the markdown file?

**Comment by alif-munim** [comment]
> Search results list different types of content - Docs, Guides, Blogs....Guides and Blog cards contain description of the content. For the Docs cards, can we display the beginning text of the markdown file?

@MelHopper this is what was done initially, but there were issues with spacing, picking up new lines, and repeating titles (this can be seen in some of the screenshots above). James and I decided it would be easier to use the `description` that's already contained in each blog, doc, guide, and news post. Some of them still need some editing, but they can definitely be changed to reflect the content of their respective markdown files.

**Comment by MelHopper** [comment]
@alif-munim Interesting point. The description tag in the markdown files has not been used effectively, so we have an issue when this tag is used, for example, for exposing within the search. I will speak to @micgibso about this, but given that the description tag has not been populated effectively for most markdown files, this is a big task to update all the topics, so in the meantime we do need a workaround where the cards simply reproduce content in the first paragraph. 

I understand the workaround is a bit fiddly - assume it will involve ignoring headers completely and simply grabbing the first piece of non-headered text? For simplicity, could we also just display all text up to the first new line to get over issues with new lines? - that should populate the card with enough copy for now.

**Comment by jcockbain** [comment]
I agree that if we are able to ignore headers and handle new lines / just use first line that'd look good. 

I like how @kimholmes had it above (https://github.com/eclipse/codewind/issues/1786#issuecomment-631637143), with the first part of the text and then '...' at the end of the content.

**Comment by alif-munim** [comment]
Hey @MelHopper and @jcockbain, thanks for the feedback! I've been looking into this further now that I'm wrapping up with the guides pages, and I think I've finally got things working how they're supposed to. Took quite a bit of fiddling with liquid filters within jekyll, but here's how paragraph excerpts look within search now:

![image](https://user-images.githubusercontent.com/59207653/84066763-68febb80-a994-11ea-85a8-267d92f99bb3.png)


One thing to note is that not all documents are laid out in exactly the same way. I've split the page content into an array and grabbed the second element as this is (usually) the non-header content. However, there are instances where this is not the case:

![image](https://user-images.githubusercontent.com/59207653/84067434-6781c300-a995-11ea-8224-cc934818cf1f.png)

Since this is only a temporary workaround though, I don't think it will be too much of an issue. Let me know what you guys think!

**Comment by MelHopper** [comment]
This is great Alif. Thank you for the information about which pages do not work with your algorithm, for being thorough in your work, this is super. It flags that we need to be more strict with certain parameters when creating new topics. A couple of questions:

1. When grabbing the topic title, are you using the LH nav tree i.e docstoc file, vs using the 1st header element in the topic? If so, are you able to instead use the 1st header element? E.g. some topics in the tree are labelled simply 'Eclipse' or 'VS Code' because they are the IDE-specific instructions that sit underneath the topic definition - see screenshot:

<img width="338" alt="Screenshot 2020-06-09 at 09 28 15" src="https://user-images.githubusercontent.com/13365734/84124811-9f3f4800-aa33-11ea-8046-0358b7bdc7d5.png">
<img width="346" alt="Screenshot 2020-06-09 at 09 28 02" src="https://user-images.githubusercontent.com/13365734/84124834-a5cdbf80-aa33-11ea-97fe-f4bb5078efdc.png">

2. I notice that the 'uninstalling Codewind for VS Code' topic does not have a description - I assume this is due to the formatting within that topic. Can you send me the markdown filename? It does not appear to be in our LH nav (another thing to highlight with the ID team).

Thanks!

**Comment by jcockbain** [comment]
@MelHopper it's https://github.com/eclipse/codewind-docs/blob/master/docs/_documentations/vsc-uninstall.md. 

Page -> https://www.eclipse.org/codewind/vsc-uninstall.html

I added a blank link between title and text on the file and it worked (for the formatting. 

It's a good point about it not being on the sidebar though, Jekyll is serving every markdown file under `/docs/documentations/` to a HTML file. In turn that is also what Jekyll stores under `site.documentations` which is what Alif is searching through. Same with news/blogs/guides. 

We have 48 md files under `_documentations` but it seems only 40 pages being linked to in `datastoc.yml`, so this will be the case for a few pages atm (built but not linked to). 

I haven't looked into this in depth but I imagine you can filter what's stored under `site.documentations` to only include the items in `datastoc.yml` https://jekyllrb.com/docs/datafiles/. Or otherwise we can ensure that we only have `md` files for pages we want to expose?

**Comment by jcockbain** [comment]
@alif-munim as Mel says I think this looks great and is close to being ready to go in. The suggestions I had were: 

1. Personal preference, and I think something that was suggested on design chat, is linked titles in the boxes, something like this: 

![image](https://user-images.githubusercontent.com/31372187/84143685-e6870200-aa4e-11ea-9a1b-afbebbd9ac7b.png)

2. Existing problem, made a bit worse by the search bar. Codewind logo overlaps the leftmost navbar link when resizing. This may be something to do outside of this PR but think it needs to collapse earlier. As you've down work on the navbar here, if you can find a way to fix would be great but otherwise could raise as a separate bug. 

![image](https://user-images.githubusercontent.com/31372187/84144006-75941a00-aa4f-11ea-9c95-be7281e932a0.png)

3. I hit this newsredirect item, when searching for news. Interestingly can't seem to get it got blogs/guides etc..: 

![image](https://user-images.githubusercontent.com/31372187/84144126-a5dbb880-aa4f-11ea-92c0-acd7c59527f8.png)

**Comment by alif-munim** [comment]
@jcockbain that's great to hear! I've definitely noticed some of these issues and agree with your suggestions. Now that most of the technical work is out of the way, I can have these edits made in no time.

As for the news redirect, I'm not sure where that's coming from but I'll look further into it and let you know if I find anything.

**Comment by alif-munim** [comment]
> I hit this newsredirect item, when searching for news. Interestingly can't seem to get it got blogs/guides etc...

I believe it's coming from the following file:
![image](https://user-images.githubusercontent.com/59207653/84161479-542a3280-aa3d-11ea-835d-55940f4d9991.png)

I could handle this file as an edge case, i.e. display a box with a link to "Latest News"
What do you think?

**Comment by jcockbain** [comment]
Ah I see what it's doing. When you hit https://www.eclipse.org/codewind/news.html it takes you to whichever is the latest news. 

For 0.14.0 we can replace with the redirect stuff being added. I'd says leave as is for now, then if https://github.com/eclipse/codewind-docs/pull/651 we can get rid of this file.

**Comment by alif-munim** [comment]
Great! Turned the headers into links and handled it as follows for now:
![image](https://user-images.githubusercontent.com/59207653/84164209-43c78700-aa40-11ea-8adf-73f38ad48e76.png)

Other than the navigation overlapping, I think this should be good to go now. I can open up a new issue for the navigation if there isn't already an existing one.

**Comment by jcockbain** [comment]
Good stuff! Yeah if you could open an issue for the header's that'd be great. 

I'd say let's wait for @MelHopper's opinion on whether we should be exposing non-sidebar docs through the search. I don't know what state some of those are in. 

The `datastoc.yaml` is what controls the sidebar contents, see this https://github.com/eclipse/codewind-docs/blob/master/docs/_includes/docs-sidebar.html#L5 line for how you can access it in the code.

**Comment by MelHopper** [comment]
Hi both, thanks for digging into this. 

1. re. _"We have 48 md files under _documentations but it seems only 40 pages being linked to in datastoc.yml, so this will be the case for a few pages atm (built but not linked to)."_

That's for the ID team to delete any files not being used for docstoc, so don't worry about changing any of the way your code is working to pick out docstoc @alif-munim . 

Agreed James - we should only have md files for topics that we want to expose. Having that markdown file title was useful because we can see where the formatting needs to be tightened up by writers. Defect raised: https://github.com/eclipse/codewind/issues/3115

2. In terms of ensuring elements in markdown files are populated and complete, I am adding to the ID team's workflow to ensure this is done. 

Cheers both!

**Comment by MelHopper** [comment]
Delivered via PR https://github.com/eclipse/codewind-docs/pull/612
